wxAuiNotebook::DoGetBestSize

walderich opened the pull request:
In my current project I encountered the issue that I needed to calculate the size of a wxAuiNotebook control (as the user can control the content of the pages during runtime using an INI file). But as `Fit` always shrinked the complete window down to size 0x0 I saw that wxAuiNotebook does not implement the `DoGetBestSize` method.

Although in my case I do not have nested/split panes (as the control gets reset after the INI file update). I thought it would make sense to still implement a generic method, so I can use this instead of writing my own, every time I need it.

The not-so-simple layout options made it a bit more complicated than I thought. But I got it to work correctly in my test cases. I did not have a deeper look into it, but currently calling `Fit` does not resize splittet tab containers. I guess this has to be handled separately.

walderich committed:
Add BestSize calculation

walderich committed:
Define MergeLayout as object member

walderich committed:
Correct merge algorithm

Use successive tests to avoid side-effects and duplicate checks.

walderich committed:
Include height of tab control

vadz commented on include/wx/aui/auibook.h:
```diff
@@ -356,6 +356,8 @@ class WXDLLIMPEXP_AUI wxAuiNotebook : public wxNavigationEnabled<wxBookCtrlBase>
     virtual bool InsertPage(size_t index, wxWindow *page, const wxString &text,
                             bool select, int imageId) wxOVERRIDE;
 
+    wxSize DoGetBestSize() const;
```
This should use `wxOVERRIDE` (not only for clarity but also because without it we're going to get clang warnings about inconsistent use of override).

walderich replied:
Done. I also added `virtual` as all other `DoGetBestSize` methods use it.

vadz commented on src/aui/auibook.cpp:
```diff
@@ -3446,6 +3446,127 @@ bool wxAuiNotebook::InsertPage(size_t index, wxWindow *page,
+    }
+};
+
+WX_DEFINE_SORTED_ARRAY(wxAuiLayoutObject*, wxAuiLayoutObjectArray);
```
Please let's avoid these macros in the new code. Just use `wxVector<>` with binary search for insertion instead.

walderich replied:
All right. Didn't know that. I guess I should do the same in my projects? But I used `wxVectorSort` instead of binary search insertion as I found [this reference](https://stackoverflow.com/questions/2710221/is-there-a-sorted-vector-class-which-supports-insert-etc/2710332#2710332) online which states that inserting in a vector might be inefficient as all later elements need to be moved, which sounds reasonable to me. I also guess that most of the layouts aren't that complicated and we won't have many vector elements anyways.

vadz commented on src/aui/auibook.cpp:
```diff
@@ -3446,6 +3446,127 @@ bool wxAuiNotebook::InsertPage(size_t index, wxWindow *page,
     }
 }
 
+class wxAuiLayoutObject
```
I'd very much appreciate a comment explaining what's going on here. AFAICS the main purpose of this class is to be able to call `MergeLayout()` on it, but I don't really know _why_ do we need to do it.

walderich replied:
I added comments to all critical parts of the algorithm.

vadz commented on src/aui/auibook.cpp:
```diff
@@ -3446,6 +3446,127 @@ bool wxAuiNotebook::InsertPage(size_t index, wxWindow *page,
+class wxAuiLayoutObject
+{
+public:
+    wxAuiLayoutObject(wxSize &s, unsigned short l, unsigned short dir, unsigned short r, unsigned short pos)
```
Maybe a comment would explain this, but do we really have to use a non-const reference here? This makes the code difficult to understand, could we just return `wxSize` from `Merge()` instead?

Also, using `layer` instead of `l`, `row` instead of `r` etc would make this ctor signature much more understandable. Alternatively, after looking at the code below, it seems like it could just take a single `wxAuiPaneInfo` parameter instead.

walderich replied:
You're right. A `const` argument is all we need here as only the copied member is changed. Returning a `wxSize` from `MergeLayout` would work, too. But I am merging the values inside the object (also multiple times, if necessary). Otherwise I would need to pass the merged `wxSize` value back to each of the `MergeLayout` calls.

vadz commented on src/aui/auibook.cpp:
```diff
@@ -3446,6 +3446,127 @@ bool wxAuiNotebook::InsertPage(size_t index, wxWindow *page,
+        row = r;
+        position = pos;
+    }
+    void MergeLayout(wxAuiLayoutObject *lo2)
```
This method should take const reference and not a non-const pointer as its argument is not supposed to be null and is not modified by it.

walderich replied:
I corrected this.

vadz commented on src/aui/auibook.cpp:
```diff
@@ -3446,6 +3446,127 @@ bool wxAuiNotebook::InsertPage(size_t index, wxWindow *page,
+        }
+    }
+
+    wxSize size;
```
Members should use `m_` prefix.

walderich replied:
Changed all members accordingly.

vadz reviewed (dismissed):
Sorry, I can't merge it in the current state because I don't really understand what does this code do and there are no comments explaining it or tests showing the expected behaviour. Could you please add them?

Also, a global style comment: could you please avoid over long (> 80 characters) lines?

TIA!

walderich committed:
Add modifiers to DoGetBestSize

Add 'wxOVERRIDE' for clarity and 'virtual' to allow further override.

walderich committed:
Pass PaneInfo to layout object

This allows to initialize all sorting data without passing them
separately to the constructor.

walderich committed:
Pass 'const' reference to MergeLayout

Ensure that reference is neither changed nor a NULL pointer.

walderich committed:
Add 'm_' prefix to class properties

walderich committed:
Replace custom sorted array by wxVector

walderich committed:
Keep PaneInfo reference to avoid copying data

walderich committed:
Correct comparison method

The method didn't take '>' into account.

walderich committed:
Rename variables to avoid overlong lines

walderich committed:
Add description to important algorithm parts

walderich committed:
Ensure that size parameter isn't changed

walderich committed:
Add newlines to avoid overlong lines

walderich commented:
Thanks for all your support and detailed explanations!

Unfortunately I didn't get the automated test running on my machine. At least not the GUI part. Therefore I tested it on my own GUI which only contains the main window, the wxAuiNotebook and a statusbar. To get the algorithm working, I added some quick & dirty `printf`s into the code and checked if it worked as expected.
Do you have an idea how I get the `test_gui` application working? I get lots of errors from the linker about missing `webkit_web_view_*` methods.

I hope I addressed all your improvements now (after my third iteration ;-) ).

vadz commented on src/aui/auibook.cpp:
```diff
@@ -3446,6 +3446,140 @@ bool wxAuiNotebook::InsertPage(size_t index, wxWindow *page,
+         * TOP/BOTTOM in vertical direction) */
+        switch (pInfo.dock_direction)
+        {
+            case wxAUI_DOCK_CENTER: m_dir = 0; break;
```
Thanks to the comment above I now understand why do we do this, but could we please have some symbolic values instead of using manifest constants? Code like `m_dir < 3` below would be more clear if it used `DockDir_Horizontal` instead.

walderich replied:
Replaced the magic numbers and used `DockDir_Vertical` for the comparisons. Although it still needs some closer look to see, why `mergeHorizontal` is used for `< DockDir_Vertical` in one place and `>= DockDir_Verical` in the other.

vadz commented on src/aui/auibook.cpp:
```diff
@@ -3446,6 +3446,140 @@ bool wxAuiNotebook::InsertPage(size_t index, wxWindow *page,
+        if (mergeHorizontal)
+        {
+            m_size.x += lo2.m_size.x;
+            if (lo2.m_size.y > m_size.y) m_size.y = lo2.m_size.y;
```
Another style nitpick: could you please put the body of this `if` on a separate line (also below)?

walderich replied:
Oh dear, I must have overlooked this.

vadz commented on src/aui/auibook.cpp:
```diff
@@ -3446,6 +3446,140 @@ bool wxAuiNotebook::InsertPage(size_t index, wxWindow *page,
+    }
+
+    wxSize m_size;
+    const wxAuiPaneInfo &m_pInfo;
```
As Travis CI failure shows, we can't use a const reference here as this class becomes neither copyable nor movable then. We can either make a copy (i.e. drop the reference part) or use a pointer, which is uglier but avoids copying. Or revert back to using individual `m_layer`, `m_dir` and `m_row` fields here, while still using `wxAuiPaneInfo` to initialize them in the ctor.

I'm leaving the choice to you, but we do need to fix the build.

walderich replied:
Of course the build must work. I corrected the reference.

vadz commented on src/aui/auibook.cpp:
```diff
@@ -3446,6 +3446,140 @@ bool wxAuiNotebook::InsertPage(size_t index, wxWindow *page,
+    size_t pos = 0;
+    for (size_t n = 1; n < layouts.size(); n++)
+    {
+        if (layouts[n].m_pInfo.dock_layer != layouts[pos].m_pInfo.dock_layer)
```
The bodies of all 3 `if`s are identical, so why can't we simplify this code to (in pseudo-code)
```
if ( SameUpToPos(layouts[n].m_pInfo, layouts[pos].m_pInfo) )
{
    layouts[pos].MergeLayout(layouts[n]);
}
else
{
    layouts[0].MergeLayout(layouts[pos]);
    pos = n;
}
```

Am I missing something here?

walderich replied:
No, your right. I simply programmed that from top to bottom and didn't recognize the equal cases.

vadz reviewed (dismissed):
Thanks for the updates but unfortunately there is still a problem (build error) and a few small comments that it'd be nice to fix. TIA!

walderich committed:
Add missing newlines after 'if' statement

walderich committed:
Simplify if-else by merging conditions

Use only one condition for all cases with equal bodies.

walderich committed:
Store PaneInfo pointer for std::move

Allow move/copy operator on LayoutObject.

walderich committed:
Use named constant for clarification

Replace magic numbers by constant to clarify checks.

walderich committed:
Create unit tests for DoGetBestSize

Create some test for verifying that wxAuiNotebook::DoGetBestSize works
as expected.

walderich committed:
Clean up unit test

Added wxDELETE to delete the temporary control from the main window
after the unit tests are completed.

walderich commented on tests/controls/auitest.cpp:
```diff
@@ -0,0 +1,137 @@
+    CHECK( auiSize.y == 275 + 4*tabHeight );
+  }
+
+  SECTION( "Surrounding panes" ) {
```
I wanted to add another test with different positions/rows. But I couldn't get it working. I tried to change the position with the `wxAuiManager`'s `AddPane` method. But this lead the test application to segfault. Does `AddPane` not work for `wxAuiNotebook` objects?

vadz reviewed (approved):
Thanks, I'm going to merge it in a moment with the changes mentioned below and also some cosmetic formatting ones.

I also had to fix a bug in `wxAuiTabContainer::Render()` to make the test pass without crashing under Windows.

walderich commented on tests/controls/auitest.cpp:
```diff
@@ -0,0 +1,137 @@
+    CHECK( auiSize.y == 175 + 3*tabHeight );
+  }
+
+  wxDELETE(nb);
```
As far as I understood the [documentation](https://github.com/catchorg/Catch2/blob/master/docs/command-line.md#aborting-after-a-certain-number-of-failures), a failed test does not stop the test case execution:

> If a REQUIRE assertion fails the test case aborts, but subsequent test cases are still run. If a CHECK assertion fails even the current test case is not aborted.

So the statements outside the `SECTION` statements can be treated like the `teardown` method in CppUnit.

vadz replied:
This is correct, but if a `REQUIRE( nb->AddPage(...) )` in the beginning of the test failed, this still would have been leaked.

walderich commented:
Thanks for the information. I think the fix is good. I didn't see any better solution.
